### PR TITLE
Drop canner dependency: datasets

### PR DESCRIPTION
**Pull request by popcornylu**
## Review note
Please update the console and graphql images.
latest test image: 
~~`ch16902-05b37de`~~
~~`ch16902-a490f2a`~~
`ch16902-32af80f`

## Test cases
List pages
- [x] List
- [x] Search
- [x] Sort
- [x] Pagination
- [x] Breadcrumb & tip link

Add/Edit pages
- [x] name, display name, description
- [x] pv auto
- [x] pv manual
- [x] env. (Please check invalid key names)
- [x] git (Please check the git ssh secret)
- [x] nfs
- [x] hostpath
- [x] Breadcrumb & tip link

Enable Upload Server
- [x] Shown only in PV, NFS
- [x] Enable and edit
- [x] Regenerate secret
- [x] Disable upload server

Group Association
- [x] GlobaL
- [x] Git, env.. Should show only readonly groups
- [x] PV, hostpath. NFS. Should show readonly & writable groups

## Screenshots
List
image

Form
image

### CODE REVIEW

File: packages/client/src/admin/Datasets/DatasetGroupsRelationTable.tsx [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+          if (!newItem) {
+            return [{id: oldItem.id}];
+          } else {
+            return [];
+          }
+        }),
+    };
+
+    if ( newValue.connect.length == 0)
+      delete newValue["connect"];
```
[the remark this answers is not in the document]
Comment (popcornylu) [reply packages/client/src/admin/Datasets/DatasetGroupsRelationTable.tsx:89]: For dataset update, there is no impact

But for dataset add, disconnect is not available in the field. This way would prevent from "disconnect" field when creating.

Query
```
mutation ($payload: DatasetCreateInput!) {
  createDataset(data: $payload) {
    id
    uploadServerSecret {
      username
      password
      __typename
    }
    __typename
  }
}
```


Variables
```
{
  "payload": {
    "name": "my-dataset",
    "displayName": "",
    "description": "",
    "mountRoot": "/datasets",
    "global": false,
    "type": "env",
    "url": "",
    "variables": {},
    "nfsServer": "",
    "nfsPath": "",
    "hostPath": "",
    "pvProvisioning": "auto",
    "volumeName": "",
    "volumeSize": 1,
    "enableUploadServer": false,
    "uploadServerLink": "",
    "groups": {
      "connect": [
        {
          "id": "d9373a4a-6103-44dc-adae-4c6a371f83b4",
          "writable": false
        }
      ],
      "disconnect": []
    }
  }
}
```

File: packages/client/src/admin/Datasets/DatasetInfo.tsx [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+          },
+        },
+      });
+
+      const secret = result?.data?.updateDataset?.uploadServerSecret;
+      if (secret) {
+        uploadServerSecretModal({
+          title: intl.formatMessage({id: 'dataset.regenerateSecretModalTitle'}),
+          secret,
+          onOk: () => {}
```
[the remark this answers is not in the document]
Comment (popcornylu) [reply packages/client/src/admin/Datasets/DatasetInfo.tsx:63]: image


No action required. Just close the dialog.

File: packages/client/src/admin/Datasets/Datasets.tsx [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+  const [queryVariables, setQueryVariables] = useState<QueryVariables>({ page: 1 });
+
+  useEffect(() => {
+    if (datasetQuery.error) {
+      errorHandler(datasetQuery.error);
+    }
+  }, [datasetQuery.error]);
+
+  if (datasetQuery.error) {
+    if (!datasetQuery?.datasetsConnection) {
```
[the remark this answers is not in the document]
Comment (popcornylu) [reply packages/client/src/admin/Datasets/Datasets.tsx:96]: If the error is from refetch (not the first fetch), we can still show the current fetched list but show the error notification at the bottom-right corner.

File: packages/client/src/admin/Datasets/Datasets.tsx [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+    return (
+      <DatasetLayout page="create">
+        <div style={styles}>
+          <DatasetForm onSubmit={handleSubmit} />
+        </div>
+      </DatasetLayout>
+    );
+  }
+
+  if (datasetQuery.loading && !datasetQuery?.datasetsConnection) {
```
[the remark this answers is not in the document]
Comment (popcornylu) [reply packages/client/src/admin/Datasets/Datasets.tsx:176]: If we remove this line, the refetch (e.g. filter, changing page) state would not show the current list.

File: packages/client/src/admin/Datasets/__tests__/DatasetInfo.spec.tsx [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+    render(
+      <TestProvider>
+        <MockedProvider mocks={mockRequests}>
+          <DatasetInfo />
+        </MockedProvider>
+      </TestProvider>
+    );
+
+    screen.debug();
```
[the remark this answers is not in the document]
Comment (popcornylu) [reply packages/client/src/admin/Datasets/__tests__/DatasetInfo.spec.tsx]: Fixed

File: packages/client/src/admin/Datasets/GitSecret.tsx [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+        </div>
+      </Modal>
+
+    </>
+  );
+}
+
+export const GitSecret = compose(
+  graphql(GetSecrets, {
+    name: 'secretsQuery',
```
[the remark this answers is not in the document]
Comment (popcornylu) [reply packages/client/src/admin/Datasets/GitSecret.tsx]: Fixed

File: packages/client/src/admin/Datasets/GitSecret.tsx [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+    options: () => {
+      return {
+        fetchPolicy: 'no-cache',
```
[the remark this answers is not in the document]
Comment (popcornylu) [reply packages/client/src/admin/Datasets/GitSecret.tsx]: Fixed\. Change `no-cache` to `network-only`

File: packages/client/src/admin/Datasets/GitSecret.tsx [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+      <Modal width={800} visible={state.modalVisible}
+        onOk={handleModalOk}
+        onCancel={handleModalCancel}>
+        <div style={{marginTop:16}}>
+          <Table
+            size='small'
+            rowKey={(data) => data.id}
+            style={{ paddingTop: 8 }}
+            columns={columns}
+            // onChange={handleTableChange}
```
[the remark this answers is not in the document]
Comment (popcornylu) [reply packages/client/src/admin/Datasets/GitSecret.tsx]: Fixed

File: packages/client/src/admin/Datasets/EnvVariables.tsx [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+  constructor(props) {
+    super(props);
+    const { value } = props;
+    this.state = {
+      fields: objectToArray(value || {}),
+    };
+  }
+
+  componentDidMount() {
+    // const {customizeValidator} = this.props;
```
[the remark this answers is not in the document]
Comment (popcornylu) [reply packages/client/src/admin/Datasets/EnvVariables.tsx]: Fixed

File: packages/client/src/admin/Datasets/DatasetInfo.tsx [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+export const DatasetInfo = compose(
+  injectIntl,
+  withRouter,
+  graphql(DatasetQuery, {
+    name: 'datasetQuery',
+    options: ({ match }: RouteComponentProps<{ id: string }>) => {
+      const datasetId = match.params.id;
+
+      return {
+        fetchPolicy: 'no-cache',
```
[the remark this answers is not in the document]
Comment (popcornylu) [reply packages/client/src/admin/Datasets/DatasetInfo.tsx]: Fixed

File: packages/client/src/admin/Datasets/DatasetGroupsRelationTable.tsx [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+  );
+}
+
+
+export const DatasetGroupsRelationTable = compose(
+  graphql(GetGroups, {
+    name: 'groupsQuery',
+    options: () => {
+      return {
+        fetchPolicy: 'no-cache',
```
[the remark this answers is not in the document]
Comment (popcornylu) [reply packages/client/src/admin/Datasets/DatasetGroupsRelationTable.tsx]: Fixed

File: packages/client/src/admin/Datasets/DatasetGroupsRelationTable.tsx [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+  color: rgba(0, 0, 0, 0.85);
+  padding: 0px 0px 8px;
+`;
+
+interface Props {
+  groups: TDatasetGroups;
+  value: TDatasetFormGroups;
+  onChange: (TDatasetFormGroups) => void;
+  allowReadOnly: boolean;
+  allowWritable: boolean;
```
[the remark this answers is not in the document]
Comment (popcornylu) [reply packages/client/src/admin/Datasets/DatasetGroupsRelationTable.tsx]: Fixed

File: packages/client/src/admin/Datasets/DatasetGroupsRelationTable.tsx [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+const Title = styled.div`
+  color: rgba(0, 0, 0, 0.85);
+  padding: 0px 0px 8px;
+`;
+
+interface Props {
+  groups: TDatasetGroups;
+  value: TDatasetFormGroups;
+  onChange: (TDatasetFormGroups) => void;
+  allowReadOnly: boolean;
```
[the remark this answers is not in the document]
Comment (popcornylu) [reply packages/client/src/admin/Datasets/DatasetGroupsRelationTable.tsx]: Fixed

File: packages/client/src/admin/Datasets/DatasetGroupsRelationTable.tsx [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+  allowWritable: boolean;
+  groupsQuery: {
+    error: Error | undefined;
+    loading: boolean;
+    groups?: any;
+    refetch;
```
[the remark this answers is not in the document]
Comment (popcornylu) [reply packages/client/src/admin/Datasets/DatasetGroupsRelationTable.tsx]: Fixed

File: packages/client/src/admin/Datasets/DatasetGroupsRelationTable.tsx [hunk 15]
```diff
@@ -0,0 +1,202 @@
+import React, { PureComponent, useState } from 'react';
```
[the remark this answers is not in the document]
Comment (popcornylu) [reply packages/client/src/admin/Datasets/DatasetGroupsRelationTable.tsx]: Fixed

File: packages/client/src/admin/Datasets/Datasets.tsx [hunk 16]
```diff
@@ -0,0 +1,354 @@
...
+                          id: dataset.node.id,
+                        },
+                      },
+                    });
+                    await datasetQuery.refetch();
+                  },
+                });
+              }}
+            >
+              {/* <Icon type="delete" /> */}
```
[the remark this answers is not in the document]
Comment (popcornylu) [reply packages/client/src/admin/Datasets/Datasets.tsx]: Removed

File: packages/client/src/admin/Datasets/Datasets.tsx [hunk 17]
```diff
@@ -0,0 +1,354 @@
...
+                        where: {
+                          id: dataset.node.id,
+                        },
+                      },
+                    });
+                    await datasetQuery.refetch();
+                  },
+                });
+              }}
+            >
```
[the remark this answers is not in the document]
Comment (popcornylu) [reply packages/client/src/admin/Datasets/Datasets.tsx]: Fixed

File: packages/client/src/admin/Datasets/Datasets.tsx [hunk 18]
```diff
@@ -0,0 +1,354 @@
...
+              {/* <Icon type="delete" /> */}
+            </Button>
```
[the remark this answers is not in the document]
Comment (popcornylu) [reply packages/client/src/admin/Datasets/Datasets.tsx]: Fixed

File: packages/client/src/admin/Datasets/EnableUploadServer.tsx [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+              id: name,
+            }
+          }}
+          onCompleted={data => {
+            const secret = get(data, 'regenerateUploadServerSecret.uploadServerSecret', null);
+            if (isPlainObject(secret) && secret.username && secret.password) {
+              uploadServerSecretModal({
+                title: intl.formatMessage({id: 'dataset.regenerateSecretModalTitle'}),
+                secret,
+                onOk: () => {}
```
[the remark this answers is not in the document]
Comment (popcornylu) [reply packages/client/src/admin/Datasets/EnableUploadServer.tsx]: Removed

File: packages/client/src/admin/Datasets/EnableUploadServer.tsx [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+import { get, isPlainObject } from 'lodash';
+import uploadServerSecretModal from './UploadServerSecretModal';
+import { RegenerateUploadServerSecretMutation } from 'queries/Datasets.graphql';
+
+interface Props {
+  intl?: any;
+  name: string;
+  allowRegenerateSecret: boolean;
+  value?: boolean;
+  onChange?;
```
[the remark this answers is not in the document]
Comment (popcornylu) [reply packages/client/src/admin/Datasets/EnableUploadServer.tsx]: Fixed

File: packages/client/src/admin/Datasets/Datasets.tsx [hunk 21]
```diff
@@ -0,0 +1,354 @@
...
+      render: (text) => text || '-',
+    },
+    {
+      key: 'type',
+      title: 'Type',
+      dataIndex: 'node.type',
+      sorter: true,
+      render: (text) => {
+        // return 'Persistent Volume';
+        if (text === 'pv') return 'Persistent Volume';
```
[the remark this answers is not in the document]
Comment (popcornylu) [reply packages/client/src/admin/Datasets/Datasets.tsx]: Changed

File: packages/client/src/admin/Datasets/Datasets.tsx [hunk 22]
```diff
@@ -0,0 +1,354 @@
...
+  datasetQuery,
+  createDatasetMutation,
+  deleteDatasetMutation,
+}: Props) {
+  const { appPrefix } = useRoutePrefix();
+  const history = useHistory();
+  const location = useLocation();
+  const querystring = new URLSearchParams(location.search);
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const [queryVariables, setQueryVariables] = useState<QueryVariables>({ page: 1 });
```
[the remark this answers is not in the document]
Comment (popcornylu) [reply packages/client/src/admin/Datasets/Datasets.tsx]: Fixed

**Review by aar0nTw — APPROVED** [review]

**Pull request merged**